provider/provider_yaml_test.go

```diff
@@ -452,14 +470,439 @@ func TestIMDSAuth(t *testing.T) {
 //
 // See https://github.com/pulumi/pulumi-aws/issues/2796
 func TestS3BucketObjectDeprecation(t *testing.T) {
+	t.Parallel()
 	ptest := pulumiTest(t, filepath.Join("test-programs", "regress-2796"), opttest.SkipInstall())
 	result := ptest.Up()
 	t.Logf("STDOUT: %v", result.StdOut)
 	t.Logf("STDERR: %v", result.StdErr)
 	require.NotContains(t, result.StdOut+result.StdErr, "aws_s3_object")
 }
 
-func configureS3() *s3sdk.Client {
+type tagsTestStep struct {
+	// The name of the Pulumi program
+	name string
+
+	// The type token of the resource, i.e. aws:s3:Bucket
+	token string
+
+	// The pulumi type of the resource, i.e. aws:s3/bucket:Bucket
+	typ string
+
+	// Constant properties for the primary resource under test.
+	//
+	// This cannot include the tags property, which will be adjusted by the test.
+	properties map[string]interface{}
+
+	// List of tags to add to the resource
+	tags map[string]interface{}
+
+	// List of default tags to add to the provider
+	defaultTags map[string]interface{}
+
+	// List of tag keys to add to the provider `ignoreTags.Keys` property
+	ignoreTagKeys []string
+
+	// Function to run prior to _any_ import step
+	preImportHook func(t *testing.T, outputs auto.OutputMap)
+
+	// Function to run after running the first up. This can be used to
+	// run extra validation
+	postUpHook func(t *testing.T, outputs auto.OutputMap)
+
+	// Other is a string that is inserted into the test program. It is intended to be
+	// used to provision supporting resources in tests.
+	other string
+
+	// If skip is non-empty, the test will be skipped with `skip` as the given reason.
+	skip string
+}
+
+// TestAccDefaultTags tries to test all the scenarios that might affect provider defaultTags / resource tags
+// i.e. up, refresh, preview, import, etc
+func TestAccDefaultTags(t *testing.T) {
+	t.Parallel()
+	if testing.Short() {
+		t.Skipf("Skipping in testing.Short() mode, assuming this is a CI run without credentials")
+	}
+
+	isNil := func(val interface{}) bool {
+		if val == nil {
+			return true
+		}
+		v, ok := val.(map[string]interface{})
+		return ok && len(v) == 0
+	}
+	validateOutputTags := func(outputs auto.OutputMap, expectedTags map[string]interface{}) {
+		stackOutputTags := outputs["actual"]
+		if !isNil(expectedTags) || !isNil(stackOutputTags) {
```

Comment:
This skips checking if one map is nil and the other isn't, perhaps also assert that to make this more complete?